test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -187,9 +194,11 @@ private static void createUI(String title, String instructions,
                                  boolean enableScreenCapture) {
         frame = new JFrame(title);
         frame.setLayout(new BorderLayout());
-        JTextArea instructionsText = new JTextArea(instructions, rows, columns);
-        instructionsText.setEditable(false);
-        instructionsText.setLineWrap(true);
+
+        JTextComponent text = instructions.startsWith("<html>")
```

Comment:
Probably `trim()` and `toLowerCase()` could be useful here.

Do we consider adding adding one more builder method to force enable html rendering?
(e.g. when we just want to highlight only few words/sentences with `<b>` without adding `<html>` tag)

Reply:
> Probably `trim()` and `toLowerCase()` could be useful here.

Could be… On the other hand, in post XHTML era, the elements are usually in lower-case. Thus we should stick with it and use lowercase HTML elements anyway.

I think `startsWith("<html>")` covers the scenario well.

Although neither `<html>` nor `<body>` are required, I prefer to always add them. These two elements don't add much overhead, do they?

> Do we consider adding adding one more builder method to force enable html rendering? (e.g. when we just want to highlight only few words/sentences with `<b>` without adding `<html>` tag)

I don't think it's needed. If you want the instructions rendered as HTML, you'll have to format them as HTML.

Forcing HTML rendering on plain text which highlights a couple of words with `<b>` will render the text unreadable: line ends are ignored by HTML, so you'll have to use `<br>` element to force line breaks. (A similar thing often happens in javadoc where a blank line is added as if to start a new paragraph, but without the explicit `<p>` element the text continues on the same line / paragraph.)

It is possible to wrap such instructions into `<html><body><pre>` but what stops you from doing it explicitly if it's what you want? (Note, however, the `<pre>` element disables line wrapping.)

I'd rather keep it simple.

Reply:
> Could be… On the other hand, in post XHTML era, the elements are usually in lower-case. Thus we should stick with it and use lowercase HTML elements anyway.

https://www.w3schools.com/html/html_elements.asp

> HTML is Not Case Sensitive
> 
> HTML tags are not case sensitive: `<P>` means the same as `<p>`.
> 
> The HTML standard does not require lowercase tags, but W3C recommends lowercase in HTML, and demands lowercase for stricter document types like XHTML.
> 
> At W3Schools we always use lowercase tag names.

So this is still a recommendation, not a requirement (we don't use XHTML).
But I agree that this can be helpful for a certain amount of consistency across tests(they can still use other tags in uppercase).


> Although neither <html> nor <body> are required, I prefer to always add them. These two elements don't add much overhead, do they?

It is a matter of personal taste. 
Normally, I prefer to indent a content of such tags by extra spaces, andThis can create additional formatting difficulties under the 80(120) character limit per line.
In those cases, I would prefer to omit those tags.

Reply:
I know HTML is not case sensitive. It's just HTML tags are always in lower case in recent years.

> But I agree that this can be helpful for a certain amount of consistency across tests(they can still use other tags in uppercase).

Yep! In a way, this will enforce `<html>` in lower case, and I hope all other tags in the instructions.



> >  Although neither nor are required, I prefer to always add them. These two elements don't add much overhead, do they?
> 
> It is a matter of personal taste.

Yes, it is… which could create inconsistencies.

> Normally, I prefer to indent a content of such tags by extra spaces, and  
> This can create additional formatting difficulties under the 80(120) character limit per line.  
> In those cases, I would prefer to omit those tags.

I don't expect too many nesting levels in instructions. However, I agree it could be problematic…

In the sample in #15661, I keep `<html><body>` and the content on the same level. Now that I looked at it again, I realised that I didn't follow the indentation consistently: first-level `<li>` aren't indented but third-level `<li>` are. I should correct this.

Because `<pre>` is used to format pieces of code, I decided not to indent the first-level `<li>` elements to avoid indentation of the code inside.

Reply:
I don't have a problem with keeping it simple. If some one writing a test doesn't see the HTML interpreted, they'll quickly be able to figure out why.
One thing about these and the other enhancements I see is that there's an (increasing) need for a nice bit of documentation in PassFailJFrame itself about all its features. 
Easiest to do this as you are adding them, so a doc comment block can be started with this one.
Probably as a class comment.

Reply:
> One thing about these and the other enhancements I see is that there's an (increasing) need for a nice bit of documentation in PassFailJFrame itself about all its features.  
> Easiest to do this as you are adding them, so a doc comment block can be started with this one.
Probably as a class comment.

Yes, I absolutely agree, the class requires adding docs. It's what I wrote in PR #15665: <q cite="https://github.com/openjdk/jdk/pull/15665#discussion_r1323068184">the entire class requires a bit of TLC for documentation.</q>

Yet I'd rather do it for PR #15665, it is easier to handle in one place rather than writing docs for one feature just to modify them right away for another feature.

Reply:
@prrace I've added description on how to use `PassFailJFrame` to #15665, it handles both HTML formatting for instructions (this PR) as well as creating test UI by passing a method reference, or rather an implementation of an interface.